File: iosApp/iosApp/views/resources/text/StringRefResolver.swift
 

import SwiftUI
import Shared


extension SharedRes.Strings {
    func asString() -> String {
        switch self {
        case .settingsThemeTitle: return "Theme"
        case .settingsThemeOptionSystem: return "System"
        case .settingsThemeOptionDark: return "Dark"
        case .settingsThemeOptionLight: return "Light"
        case .settingsThemeDialogTitle: return "Choose Theme"
        case .settingsFavoriteSwipeTitle: return "Swipe to remove"
        case .settingsFavoriteSwipeSummaryOn: return "Swipe enabled"
        case .settingsFavoriteSwipeSummaryOff: return "Swipe disabled"
        case .settingsSyncTitle: return "Content updates"
        case .settingsSyncSummary: return "Keep app content up to date"
        case .settingsSyncInProgress: return "Updating content…"
        case .settingsCategoryInterface: return "Interface"
        case .settingsCategoryData: return "Data"
        case .settingsAppversionTitle: return "App Version"
        case .settingsAppversionSummary: return "Version info"
        case .settingsDeviceinfoTitle: return "Device Info"
        case .settingsSyncDialogTitle: return "Check for updates?"
        case .settingsSyncDialogMessage: return "\nWe’ll refresh the app content with the latest available data.\n"
        default: return String(describing: self)
        }
    }
}


extension StringRef {
    func asString() -> String {
        if let ref = self as? SharedRes.__Strings {
            return ref.toSwiftEnum().asString()
        }
        return String(describing: self)
    }
}


extension StringRefWithArgs {
    func asStringWithArg(_ arg: String) -> String {
        switch self.ref {
        case SharedRes.Strings.settingsSummaryCurrentFmt: return "Current: \(arg)"
        case SharedRes.Strings.settingsSyncLastFmt: return "Last updated: \(arg)"
        case SharedRes.Strings.settingsSyncSuccessFmt: return "Content updated successfully\nLast updated: \(arg)"
        case SharedRes.Strings.settingsSyncFailedFmt: return "Update failed. Please try again.\nLast updated: \(arg)"
        default: return self.ref.simpleName()
        }
    }
}

File: composeApp/src/commonMain/kotlin/com/exploramus/app/resources/StringRefResolver.kt
package com.exploramus.app.resources

import androidx.compose.runtime.Composable
import com.exploramus.shared.resources.SharedRes
import com.exploramus.shared.resources.StringRef
import com.exploramus.shared.resources.StringRefWithArgs
import exploramus.composeapp.generated.resources.*
import org.jetbrains.compose.resources.stringResource

object StringRefResolver {
    @Composable
    fun resolve(ref: StringRef): String = when (ref) {
        SharedRes.Strings.settings_theme_title -> stringResource(Res.string.settings_theme_title)
        SharedRes.Strings.settings_theme_option_system -> stringResource(Res.string.settings_theme_option_system)
        SharedRes.Strings.settings_theme_option_dark -> stringResource(Res.string.settings_theme_option_dark)
        SharedRes.Strings.settings_theme_option_light -> stringResource(Res.string.settings_theme_option_light)
        SharedRes.Strings.settings_theme_dialog_title -> stringResource(Res.string.settings_theme_dialog_title)
        SharedRes.Strings.settings_favorite_swipe_title -> stringResource(Res.string.settings_favorite_swipe_title)
        SharedRes.Strings.settings_favorite_swipe_summaryOn -> stringResource(Res.string.settings_favorite_swipe_summaryOn)
        SharedRes.Strings.settings_favorite_swipe_summaryOff -> stringResource(Res.string.settings_favorite_swipe_summaryOff)
        SharedRes.Strings.settings_sync_title -> stringResource(Res.string.settings_sync_title)
        SharedRes.Strings.settings_sync_summary -> stringResource(Res.string.settings_sync_summary)
        SharedRes.Strings.settings_sync_dialog_title -> stringResource(Res.string.settings_sync_dialog_title)
        SharedRes.Strings.settings_sync_dialog_message -> stringResource(Res.string.settings_sync_dialog_message)
        SharedRes.Strings.settings_category_interface -> stringResource(Res.string.settings_category_interface)
        SharedRes.Strings.settings_category_data -> stringResource(Res.string.settings_category_data)
        SharedRes.Strings.settings_appversion_title -> stringResource(Res.string.settings_appversion_title)
        SharedRes.Strings.settings_appversion_summary -> stringResource(Res.string.settings_appversion_summary)
        SharedRes.Strings.settings_deviceinfo_title -> stringResource(Res.string.settings_deviceinfo_title)
        SharedRes.Strings.settings_deviceinfo_summary -> stringResource(Res.string.settings_deviceinfo_summary)
        SharedRes.Strings.settings_sync_in_progress -> stringResource(Res.string.settings_sync_in_progress)
        else -> ref.simpleName()
    }

    @Composable
    fun resolve(ref: StringRefWithArgs, vararg args: String): String = when (ref.ref) {
        SharedRes.Strings.settings_summary_current_fmt ->
            stringResource(Res.string.settings_summary_current, *args)
        else -> ref.ref.simpleName()
    }

    @Composable
    fun resolve(ref: StringRefWithArgs, arg: String): String = when (ref.ref) {
        SharedRes.Strings.settings_sync_last_fmt ->
            stringResource(Res.string.settings_sync_last, arg)

        SharedRes.Strings.settings_sync_success_fmt ->
            stringResource(Res.string.settings_sync_success, arg)

        SharedRes.Strings.settings_sync_failed_fmt ->
            stringResource(Res.string.settings_sync_failed, arg)

        else -> ref.ref.simpleName()
    }
}
